### PR TITLE
Feature/sha1-sha256-md5 default search/download/rescan

### PR DESCRIPTION
**Pull request by JavierBotella**

### CODE REVIEW

File: src/polyswarm_api/__init__.py [hunk 1]
```diff
@@ -556,16 +575,25 @@ def _fix_result(self, result):
...
         Download a file via the PS API
 
         :param h: Hash of the file you wish to download
         :param destination_dir: Directory you wish to save the file in
         :param with_metadata: Whether to save related file metadata into an associated JSON file
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
         :return: Dictionary containing path to the downloaded file if successful, error message if not
         """
+        if not hash_type or hash_type == 'sha1' or \
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: Create `supported_hashtypes` attribute for this class, which is a list of hash types. Then just check `if not hash_type in self.supported_hashtypes`

File: src/polyswarm_api/__init__.py [hunk 2]
```diff
@@ -451,14 +461,23 @@ def _fix_result(self, result):
...
+    async def rescan_hash(self, to_rescan, hash_type=None):
         """
         Start a rescan for single hash using the PS API asynchronously.
 
         :param to_rescan: hash of the file to rescan
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
         :return: JSON report file
         """
+        if not hash_type or hash_type == 'sha1' or \
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: Same thing here, we shouldn't have to add checks to this every time we add a hash

File: src/polyswarm_api/__init__.py [hunk 3]
```diff
@@ -372,14 +373,23 @@ def _fix_result(self, result):
...
+    async def search_hash(self, to_scan, hash_type=None):
         """
         Search for a single hash using the PS API asynchronously.
 
         :param to_scan: Hash to search for
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
         :return: JSON report file
         """
+        if not hash_type or hash_type == 'sha1' or \
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: Same here

File: src/polyswarm_api/__main__.py [hunk 4]
```diff
@@ -227,45 +177,32 @@ def search():
 
 
 @click.option('-r', '--hash-file', help='File of hashes, one per line.', type=click.File('r'))
-@click.option('--hash-type', help='Hash type to search [sha256|sha1|md5], default=sha256', default='sha256')
+@click.option('--hash-type', help='Hash type to search [default:sha256|sha1|md5]', default=None)
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: default is now "autodetect"

File: src/polyswarm_api/__main__.py [hunk 5]
```diff
@@ -227,45 +177,32 @@ def search():
...
     api = ctx.obj['api']
 
     hashes = list(hashes)
 
     if hash_file:
-        hashes += _get_hashes_from_file(hash_file)
+        hashes += get_hashes_from_file(hash_file)
+
+    # validate default hashes [sha1|sha256|md5]
+    if not hash_type or hash_type == 'sha1' or \
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: Same thing, there should be a global list of supported hashes somewhere so we don't have to change 5 different `if` statements if we update it.

File: src/polyswarm_api/__main__.py [hunk 6]
```diff
@@ -332,29 +269,29 @@ def lookup(ctx, uuid, uuid_file):
 
 @click.option('-r', '--hash-file', help='File of hashes, one per line.', type=click.File('r'))
 @click.option('-m', '--metadata', is_flag=True, default=False, help='Save file metadata into associated JSON file')
-@click.option('--hash-type', help='Hash type to search [sha256|sha1|md5], default=sha256', default='sha256')
+@click.option('--hash-type', help='Hash type to search [default:sha256|sha1|md5]', default=None)
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: default autodetect

File: src/polyswarm_api/__main__.py [hunk 7]
```diff
@@ -332,29 +269,29 @@ def lookup(ctx, uuid, uuid_file):
...
 @click.argument('hash', 'hash', nargs=-1, callback=validate_hash)
 @click.argument('destination', 'destination', nargs=1, type=click.Path(file_okay=False))
 @polyswarm.command('download', short_help='download file(s)')
 @click.pass_context
 def download(ctx, metadata, hash_file, hash_type, hash, destination):
-    if not os.path.exists(destination):
-        os.makedirs(destination)
-
+    """
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: When adding docstrings, please document the parameters as well. There's a backlog of these but as we are doing it, good to do it right.

File: src/polyswarm_api/__main__.py [hunk 8]
```diff
@@ -363,25 +300,25 @@ def download(ctx, metadata, hash_file, hash_type, hash, destination):
 
 
 @click.option('-r', '--hash-file', help='File of hashes, one per line.', type=click.File('r'))
-@click.option('--hash-type', help='Hash type to search [sha256|sha1|md5], default=sha256', default='sha256')
+@click.option('--hash-type', help='Hash type to search [default:sha256|sha1|md5]', default=None)
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: default autodetect

File: src/polyswarm_api/__main__.py [hunk 9]
```diff
@@ -363,25 +300,25 @@ def download(ctx, metadata, hash_file, hash_type, hash, destination):
...
 @click.argument('hash', 'hash', nargs=-1, callback=validate_hash)
 @polyswarm.command('rescan', short_help='rescan files(s) by hash')
 @click.pass_context
 def rescan(ctx, hash_file, hash_type, hash):
+    """
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: Document params

File: src/polyswarm_api/__main__.py [hunk 10]
```diff
@@ -363,25 +300,25 @@ def download(ctx, metadata, hash_file, hash_type, hash, destination):
...
-            if (hash_type == 'sha256' and _is_valid_sha256(h)) or \
-                    (hash_type == 'sha1' and _is_valid_sha1(h)) or \
-                    (hash_type == 'md5' and _is_valid_md5(h)):
-                hashes.append(h)
-            else:
-                logger.warning('Invalid hash %s in file, ignoring.', h)
+        hashes += get_hashes_from_file(hash_file)
+
+    # validate default hashes [sha1|sha256|md5]
+    if not hash_type or hash_type == 'sha1' or \
```
Comment (supernothing) [inline src/polyswarm_api/__main__.py]: Fix `if` statement

File: src/polyswarm_api/__init__.py [hunk 11]
```diff
@@ -372,14 +373,23 @@ def _fix_result(self, result):
...
             return await self.scan_fileobj(fobj, os.path.basename(to_scan))
 
-    async def search_hash(self, to_scan, hash_type='sha256'):
+    async def search_hash(self, to_scan, hash_type=None):
         """
         Search for a single hash using the PS API asynchronously.
 
         :param to_scan: Hash to search for
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: default is autodetect

File: src/polyswarm_api/__init__.py [hunk 12]
```diff
@@ -451,14 +461,23 @@ def _fix_result(self, result):
...
         return response
 
-    async def rescan_hash(self, to_rescan, hash_type='sha256'):
+    async def rescan_hash(self, to_rescan, hash_type=None):
         """
         Start a rescan for single hash using the PS API asynchronously.
 
         :param to_rescan: hash of the file to rescan
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: default is autodetect

File: src/polyswarm_api/__init__.py [hunk 13]
```diff
@@ -556,16 +575,25 @@ def _fix_result(self, result):
...
-    async def download_file(self, h, destination_dir, with_metadata=False, hash_type='sha256'):
+    async def download_file(self, h, destination_dir, with_metadata=False, hash_type=None):
         """
         Download a file via the PS API
 
         :param h: Hash of the file you wish to download
         :param destination_dir: Directory you wish to save the file in
         :param with_metadata: Whether to save related file metadata into an associated JSON file
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: default is autodetect

File: src/polyswarm_api/__init__.py [hunk 14]
```diff
@@ -586,14 +614,14 @@ def _fix_result(self, result):
...
-    async def download_files(self, hashes, destination_dir, with_metadata=True, hash_type='sha256'):
+    async def download_files(self, hashes, destination_dir, with_metadata=True, hash_type=None):
         """
         Download files  via the PS API
 
         :param hashes: Hashes of the files you wish to download
         :param destination_dir: Directory you wish to save the files in
         :param with_metadata: Whether to save related file metadata into an associated JSON files
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [default:sha256|sha1|md5]
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: autodetect. update the other docstrings too

File: src/polyswarm_api/__main__.py (hunk 5 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: yep. i was thinking about same thing, but we'll have to rewrite it with the non defaults ones. Anyway i will do something with this.

File: src/polyswarm_api/__main__.py (hunk 4 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: this is for the if blocks

Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: done

File: src/polyswarm_api/__init__.py (hunk 11 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__init__.py (hunk 12 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__init__.py (hunk 13 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__init__.py (hunk 14 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__main__.py (hunk 6 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: done

File: src/polyswarm_api/__main__.py (hunk 8 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: done

File: src/polyswarm_api/__main__.py (hunk 7 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: yep, it's missing in all the funcs around.

Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: we should document all at once in next PR.

File: src/polyswarm_api/__init__.py (hunk 1 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__init__.py (hunk 2 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__init__.py (hunk 3 above)
Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

File: src/polyswarm_api/__main__.py (hunk 5 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: done

File: src/polyswarm_api/__main__.py (hunk 10 above)
Comment (JavierBotella) [reply src/polyswarm_api/__main__.py]: done

File: src/polyswarm_api/__init__.py [hunk 15]
```diff
@@ -372,14 +373,20 @@ def _fix_result(self, result):
...
             return await self.scan_fileobj(fobj, os.path.basename(to_scan))
 
-    async def search_hash(self, to_scan, hash_type='sha256'):
+    async def search_hash(self, to_scan, hash_type=None):
         """
         Search for a single hash using the PS API asynchronously.
 
         :param to_scan: Hash to search for
-        :param hash_type: Hash type [sha256|sha1|md5]
+        :param hash_type: Hash type [autodetect:sha256|sha1|md5]
```
Comment (supernothing) [inline src/polyswarm_api/__init__.py]: I wasn't clear the first time, this should be `:param hash_type: Hash type [default:autodetect, sha256|sha1|md5]`

Comment (JavierBotella) [reply src/polyswarm_api/__init__.py]: done

**Pull request merged**